R/Step1_QUAIL_rank_score.R

```diff
@@ -0,0 +1,182 @@
+  
+  # Fit quantile regression
+  Qreg <- suppressWarnings(rq(pheno ~ ., data = covar_data, tau = tau, method = "fn"))
+  
+  # Get rank score
+  coeff <- summary(Qreg, se = "ker")$coefficients
+  SE_d_rv <- coeff[nrow(coeff), 2]
+  a_i_tau <- (tau - ifelse(residuals(Qreg) < 0, 1, 0)) * SE_d_rv / sqrt(-tau^2 + tau)
+  
+  return(a_i_tau)
+}
+
+#' Fit Rank Scores for All Quantile Levels
+#' @param pheno numeric vector of phenotype values
+#' @param covariates matrix/data.frame of covariates
+#' @param num_tau_levels integer number of quantile levels
+#' @param num_cores integer number of cores for parallel processing
+#' @return list of rank scores for each quantile level
+#' @importFrom parallel mclapply
+fit_rank_scores <- function(pheno, covariates, num_tau_levels, num_cores = 1) {
+  if (num_cores > 1 && requireNamespace("parallel", quietly = TRUE)) {
+    mclapply(1:num_tau_levels, function(i) {
+      tau <- i / (num_tau_levels + 1)
+      calculate_rank_score(pheno, covariates, tau)
+    }, mc.cores = num_cores)
+  } else {
+    lapply(1:num_tau_levels, function(i) {
+      tau <- i / (num_tau_levels + 1)
+      calculate_rank_score(pheno, covariates, tau)
+    })
+  }
+}
+
+#' Calculate Integrated Rank Score
+#' @param rank_scores list of rank scores
+#' @param method character "equal" or "ivw"
+#' @param num_tau_levels integer number of quantile levels
+#' @return vector of integrated rank scores
+#' @importFrom quadprog solve.QP
+#' @importFrom stats cov
+calculate_integrated_score <- function(rank_scores, method = "equal", num_tau_levels) {
+  if (method == "equal") {
+    int_rank_score <- 0
+    for (i in 1:num_tau_levels) {
+      weight <- ifelse((i > (num_tau_levels) / 2), 1, -1)
+      int_rank_score <- int_rank_score + weight * rank_scores[[i]]
+    }
+  } else if (method == "ivw") {
+    # IVW method implementation
+    a_i_tau_diff_matrix <- matrix(0, 
+                                nrow = length(rank_scores[[1]]), 
+                                ncol = num_tau_levels / 2)
+    for (i in 1:(num_tau_levels / 2)) {
+      a_i_tau_diff_matrix[, i] <- rank_scores[[i + num_tau_levels / 2]] - rank_scores[[i]]
+    }
+    
+    # Optimization steps
+    Y_QI_Var_Cov <- cov(a_i_tau_diff_matrix)
+    Dmat <- Y_QI_Var_Cov
+    dvec <- rep(0, num_tau_levels / 2)
+    Amat <- cbind(rep(1, num_tau_levels / 2), diag(num_tau_levels / 2))
+    bvec <- c(1, rep(0, num_tau_levels / 2))
+    
+    qp <- solve.QP(Dmat, dvec, t(Amat), bvec, meq = 1)
+    weights_vec <- qp$solution
+    
+    int_rank_score <- 0
+    for (i in 1:(num_tau_levels / 2)) {
+      int_rank_score <- int_rank_score + weights_vec[i] * 
+        (rank_scores[[i + num_tau_levels / 2]] - rank_scores[[i]])
+    }
+  }
+  
+  return(int_rank_score / (num_tau_levels / 2))
+}
+
+### calculate_integrated_score considering odds and even number tau
+calculate_integrated_score <- function(rank_scores, method = "equal", num_tau_levels) {
```

Comment:
@al4225 You have this same function defined twice with different implementations. I take that you want ot keep the latter? Either way please remove / merge.